### PR TITLE
Sign commits

### PR DESCRIPTION
**Pull request by jdkandersson**
Makes commits signed when required by a given repository.

This requires an e2e tests, they are still being worked on in #183. In the meantime, a successful run is available here: https://github.com/canonical/wordpress-k8s-operator/pull/96

### CODE REVIEW

File: src/repository.py [hunk 1]
```diff
@@ -342,15 +396,37 @@ def update_branch(
...
             Repository client with the updated branch
         """
+        push_args = ["-u"]
+        if force:
+            push_args.append("-f")
+        push_args.extend([ORIGIN_NAME, self.current_branch])
+
         try:
+            # Create the branch if it doesn't exist
+            if push:
```
Comment (arturo-seijas) [inline src/repository.py:406]: What's the points of pushing before and after commiting?

File: src/repository.py [hunk 2]
```diff
@@ -342,15 +396,37 @@ def update_branch(
...
-                    args.append("-f")
-                args.extend([ORIGIN_NAME, self.current_branch])
-                self._git_repo.git.push(*args)
+                try:
+                    self._git_repo.git.push(*push_args)
+                except GitCommandError as exc:
+                    # Try with the PyGithub client, suppress any errors and report the original
+                    # problem on failure
+                    try:
+                        logging.info(
```
Comment (arturo-seijas) [inline src/repository.py:418]: I think it might be worth to print the exception msg here

File: src/repository.py (hunk 1 above)
Comment (jdkandersson) [reply src/repository.py:406]: To create the branch, the PyGithub API doesn't allow for creating branches from specific commits, with a force git push it is allowed. Without pushing the branch first, GitHub actually allows unsigned commits through. By first pushing to create the branch and then pushing the commit, we catch that case and sign the commits in that case too

File: tests/unit/test_repository.py [hunk 3]
```diff
@@ -838,30 +861,146 @@ def side_effect(*args):
...
+
+    repository_client._github_client_push(commit_files=(commit_file,), commit_msg=commit_msg)
+
+    mock_github_repo.get_branch.assert_called_once_with(DEFAULT_BRANCH)
+    mock_github_repo.get_git_tree.assert_called_once_with(
+        sha=mock_github_repo.get_branch.return_value.commit.sha
+    )
+
+    # The InputGitTreeElement does not expose its arguments, can only check that the correct number
+    # were passed
```
Comment (gregory-schiano) [inline tests/unit/test_repository.py:884]: Same comment as above

File: src/repository.py [hunk 4]
```diff
@@ -133,6 +135,39 @@ def __str__(self) -> str:
...
+        case commit_module.FileAdded:
+            commit_file = cast(commit_module.FileAdded, commit_file)
+            return InputGitTreeElement(
+                path=str(commit_file.path), mode="100644", type="blob", content=commit_file.content
+            )
+        case commit_module.FileModified:
+            commit_file = cast(commit_module.FileModified, commit_file)
+            return InputGitTreeElement(
+                path=str(commit_file.path), mode="100644", type="blob", content=commit_file.content
+            )
```
[the remark this answers is not in the document]
Comment (jdkandersson) [reply src/repository.py:160]: yeh I think modified should just be mapped as added, before using the git tree there was a difference but now there isn't a good reason to keep them separate

[the remark this answers is not in the document]
Comment (jdkandersson) [reply src/repository.py:160]: should be fixed in next push

File: tests/unit/test_repository.py [hunk 5]
```diff
@@ -838,30 +861,146 @@ def side_effect(*args):
...
+
+    (repository_path / DOCUMENTATION_FOLDER_NAME).mkdir()
+    (repository_path / (file_path := Path(f"{DOCUMENTATION_FOLDER_NAME}/test.text"))).write_text(
+        "content 1", encoding="utf-8"
+    )
+
+    mock_git_repository = mock.MagicMock(spec=Git)
+    mock_git_repository.add = mock.Mock(return_value=None)
+    mock_git_repository.commit = mock.Mock(return_value=None)
+    mock_git_repository.show = mock.Mock(return_value=f"A {file_path}")
```
[the remark this answers is not in the document]
Comment (jdkandersson) [reply tests/unit/test_repository.py:970]: ah yes, thank you! Should be solved in next commit

File: src/repository.py (hunk 1 above)
Comment (arturo-seijas) [reply src/repository.py:401]: Thx for the explanation

File: tests/unit/test_repository.py [hunk 6]
```diff
@@ -34,6 +35,28 @@
...
+)
+def test__commit_file_to_tree_element(commit_file: commit.FileAdded):
+    """
+    arrange: given commit file
+    act: when _commit_file_to_tree_element is called with the commit file
+    assert: then a InputGitTreeElement is returned.
+    """
+    tree_element = repository._commit_file_to_tree_element(commit_file=commit_file)
+
+    # InputGitTreeElement don't expose any data, can only check that the correct type is returned
```
[the remark this answers is not in the document]
Comment (jdkandersson) [reply tests/unit/test_repository.py:55]: thinking about this, it is probably ok to use that, let me try

[the remark this answers is not in the document]
Comment (jdkandersson) [reply tests/unit/test_repository.py:55]: should be fixed in next push

File: tests/unit/test_repository.py (hunk 3 above)
Comment (jdkandersson) [reply tests/unit/test_repository.py:883]: should be fixed on next push

**Review by gregory-schiano — APPROVED** [review]
LGTM

**Review by yanksyoon — APPROVED** [review]
LGTM!

**Pull request merged**